File: LeetCodePractice/src/Greedy/LongestHappyString.java
package Greedy;
/*
        A string is called happy if it does not have any of the strings 'aaa', 'bbb' or 'ccc' as a substring.
        Given three integers a, b and c, return any string s, which satisfies following conditions:

        s is happy and longest possible.
        s contains at most a occurrences of the letter 'a', at most b occurrences of the letter 'b' and at most c occurrences of the letter 'c'.
        s will only contain 'a', 'b' and 'c' letters.
        If there is no such string s return the empty string "".

        Example 1:
        Input: a = 1, b = 1, c = 7
        Output: "ccaccbcc"
        Explanation: "ccbccacc" would also be a correct answer.

        Example 2:
        Input: a = 2, b = 2, c = 1
        Output: "aabbc"

        Example 3:
        Input: a = 7, b = 1, c = 0
        Output: "aabaa"
        Explanation: It's the only correct answer in this case.

        Constraints:
        0 <= a, b, c <= 100
        a + b + c > 0

Intuition: this is almost identical to 984. String Without AAA or BBB. We just need to ignore the smallest count in each round.
Aassuming a >= b >= c: always try to add 'aa'. If a - 2 >= b, add 'b' (if not, the next round will add 'bb'). Repeat recursivelly for the remaining counts.

In other words, we are greedily use two characters from the largest pile. We cusion these two characters with a character from the medium pile.
For [11,5,3], as an example, we first generate 'aabaabaab', and our piles become [5,2,3]. At this time, c becomes the medium pile, and we generate '..aac' ([3,2,2]).
After we add one more '..aa', c becomes the largest pile and we pull two characters from it '..cc' ([1,2,0]). We add the rest '..bba', and the resulting string is 'aabaabaabaacaaccbba'.
*/
public class LongestHappyString {
    public String longestDiverseString(int a, int b, int c) {
        return dfs(a, b, c, "a", "b", "c");
    }

    private String dfs(int a, int b, int c, String as, String bs, String cs){
        if(a < b){
            return dfs(b, a, c, bs, as, cs);
        }
        if(b < c){
            return dfs(a, c, b, as, cs, bs);
        }
        // make sure order a >= b >= c
        StringBuilder sb = new StringBuilder();
        int a_use = Math.min(a, 2);
        for(int i = 0; i < a_use; i++){
            sb.append(as);
        }
        if(b == 0){
            return sb.toString();
        }
        int b_use = a - a_use >= b ? 1 : 0;
        for(int i = 0; i < b_use; i++){
            sb.append(bs);
        }
        sb.append(dfs(a - a_use, b - b_use, c, as, bs, cs));
        return sb.toString();
    }
}

File: LeetCodePython/LongestHappyString.py
"""
A string s is called happy if it satisfies the following conditions:

s only contains the letters 'a', 'b', and 'c'.
s does not contain any of "aaa", "bbb", or "ccc" as a substring.
s contains at most a occurrences of the letter 'a'.
s contains at most b occurrences of the letter 'b'.
s contains at most c occurrences of the letter 'c'.
Given three integers a, b, and c, return the longest possible happy string. If there are multiple longest happy strings, return any of them. If there is no such string, return the empty string "".

A substring is a contiguous sequence of characters within a string.

Example 1:
Input: a = 1, b = 1, c = 7
Output: "ccaccbcc"
Explanation: "ccbccacc" would also be a correct answer.

Example 2:
Input: a = 7, b = 1, c = 0
Output: "aabaa"
Explanation: It is the only correct answer in this case.

Constraints:
0 <= a, b, c <= 100
a + b + c > 0

hints:
1 Use a greedy approach.
2 Use the letter with the maximum current limit that can be added without breaking the condition.
"""

class LongestHappyString:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        curra, currb, currc = 0, 0, 0
        # Maximum total iterations possible is given by the sum of a, b, and c.
        total_iterations = a + b + c
        result = []

        for i in range(total_iterations):
            if (a >= b and a >= c and curra != 2) or (
                    a > 0 and (currb == 2 or currc == 2)
            ):
                # If 'a' is maximum and its streak is less than 2, or if streak of 'b' or 'c' is 2, then 'a' will be the next character.
                result.append("a")
                a -= 1
                curra += 1
                currb = 0
                currc = 0
            elif (b >= a and b >= c and currb != 2) or (
                    b > 0 and (currc == 2 or curra == 2)
            ):
                # If 'b' is maximum and its streak is less than 2, or if streak of 'a' or 'c' is 2, then 'b' will be the next character.
                result.append("b")
                b -= 1
                currb += 1
                curra = 0
                currc = 0
            elif (c >= a and c >= b and currc != 2) or (
                    c > 0 and (curra == 2 or currb == 2)
            ):
                # If 'c' is maximum and its streak is less than 2, or if streak of 'a' or 'b' is 2, then 'c' will be the next character.
                result.append("c")
                c -= 1
                currc += 1
                curra = 0
                currb = 0

        return "".join(result)